changefeedccl: handle changefeed replan with MuxRangefeed; default MuxRangefeed to true

miretskiy opened the pull request:
Flip `changefeed.mux_rangefeed.enabled` to true to have changefeeds use MuxRangefeed RPC.

Fixes #95781

Release note (enterprise change): Changefeed use MuxRangefeed RPC, which improves changefeed scalability.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95822)
<!-- Reviewable:end -->

erikgrinaker commented:
@miretskiy Didn't we see issues with this when we enabled it in production?

miretskiy commented:
we are trying to work through some of the issues;

Commit:
kvcoord: Fix multiple MuxRangeFeed client bugs

This PR addresses multiple issues with MuxRangeFeed client
implementation, identified by running various CDC stress tests.

The first issue was incorrect context used when establishing
MuxRangeFeed with a node.  When calling `startMuxRangeFeed`, the caller
provided context was used to establish the RPC with a node.
However, this context was meant to be used only for a single range
rangefeed -- and not for a node level MuxRangeFeed RPC.  As a result,
as soon as the first rangefeed completed (perhaps due to an error, such
as range split), the entire node level MuxRangeFeed, along with all of
the ranges running on that node, would be canceled.

The second issue is a race between consumer termination (consumer is a
single range rangefeed), and the "demux" loop.  There were really two
races here.  First, when we return an error to the caller (e.g. range
split), the caller simply executes cleanup and returns.  The cleanup
removes the information about the stream from the mux rangefeed, which
produces an error.  Secondly, it is possible for the demux loop to
observe additional events even after receving event error.
This is because the rangefeed processor (server side) termination is asynchronous.
It is possible that when the server shuts down (i.e. cancels) output
loop due to an error (range split), the internal buffer may still have
some events (such as checkpoint).  It's a race whether or not such
event will be sent on the RPC stream.  If such event arrives,
demux loop will block forever because the caller already terminated.

The fixes in thisThis were verified by verifying clean, non-flaky
execution over 25000 runs of multiple CDC tests
(TestChangefeedInitialScanOnly, and others).

Informs #95781

Release note: None

Commit:
changefeedccl: Metamorphically enable mux range feed

Allow `changefeed.mux_rangefeed.enabled` to be metamorphically
set in tests.  Note: the default still remains false.

Informs #95781

Epic: None

Release note: None
